tests/mesh/mesh_assign.C

```diff
@@ -0,0 +1,230 @@
+       else
+       {
+        libmesh_error_msg("Error: specified mesh_type not understood");
+       }
+
+       MeshTools::Generation::build_square(*mesh_one,
+                                          2, 2,
+                                          0., 1.,
+                                          0., 1.,
+                                          QUAD9);
+
+       std::cout<<"Info from mesh one."<<std::endl<<std::endl;
+       mesh_one->print_info();
+
+       // build_square adds boundary_ids 0,1,2,3 for the bottom, right,
+       // top, and left sides, respectively.
+
+       /** Create an ES based on the mesh, with a single variable.
+        * We will set non-homogenous bc u = 1.0 on the bottom and right
+        * sides of the mesh. After moving a new mesh into mesh_one, we will
+        * verify that u is projected properly onto the new mesh_one topology.
+        */
+       EquationSystems es(*mesh_one);
+       System & system = es.add_system<System> ("SimpleSystem");
+       system.add_variable("u", FIRST);
+
+       // Now set the Dirichlet bcs: u = 1.0, when boundary id = 0 and 1.
+       // The function describing the boundary value
+       ConstFunction<Number> one(1.0);
+
+       // The boundary ids of the boundaries where the value is to be set.
+       const boundary_id_type all_bdry_ids[4] = {0, 1, 2, 3};
+       const std::set<boundary_id_type> bottom_right_bdry_ids(all_bdry_ids, all_bdry_ids + 1);
+
+       // The variables for which we are defining the bc.
+       std::vector<unsigned int> variables;
+       variables.push_back(0);
+
+       DirichletBoundary dirichlet_bc(bottom_right_bdry_ids, variables, &one);
+
+       system.get_dof_map().add_dirichlet_boundary(dirichlet_bc);
+
+       es.init();
+
+       if (mesh_type.compare("Mesh") == 0)
+       {
+        mesh_two = std::make_shared<Mesh>(*TestCommWorld);
+       }
+       else if (mesh_type.compare("replicated") == 0)
+       {
+        mesh_two = std::make_shared<ReplicatedMesh>(*TestCommWorld);
+       }
+       else if (mesh_type.compare("distributed") == 0)
+       {
+        mesh_two = std::make_shared<DistributedMesh>(*TestCommWorld);
+       }
+       else
+       {
+        libmesh_error_msg("Error: specified mesh_type not understood");
+       }
+
+       if(mesh_two_case_type.compare("from_memory") == 0)
+       {
+         /**
+          * Build another 2d 2x2 square mesh (mesh_two) covering [0.0, 1.0] x [0.0, 1.0]
+          * with linear Quad elements. We will later refine this mesh and replace
+          * mesh_one with it.
+          */
+
+          MeshTools::Generation::build_square(*mesh_two,
+                                              2, 2,
+                                              0., 1.,
+                                              0., 1.,
+                                              QUAD4);
+
+          // Uniformly refine mesh_two
+          std::unique_ptr<MeshRefinement> mesh_refinement(new MeshRefinement(*mesh_two));
```

Comment:
`libmesh_make_unique` should be used instead of `new`. Basically if you find yourself writing `new Foo`, you should stop and figure out whether it's really necessary. My goal is to have as few raw new/delete in the code as possible.

Reply:
I was just beginning to get the hang of new ....

Reply:
I had to help some users upgrade a MOOSE app last month to add DistributedMesh capability.  A MOOSE app that, to better match the style of a different application they were coupling to, used malloc().

It was difficult to make only the necessary updates and resist the temptation to prescribe a more deep-rooted solution.

https://www.youtube.com/watch?v=cDfqwHck2Qg

Reply:
Basics are always the best.